src/Psl/Option/Option.php

```diff
@@ -0,0 +1,260 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Option;
+
+use Closure;
+
+/**
+ * @template T
```

Comment:
it should be `@template-covariant` to accept subtypes. That's really useful in real applications: for example a function accepts `Option<Organization>`,  and you'd like to give it `Option<GovernmentOrganization>`.

Reply:
also, `@immutable` / `@psalm-immutable` is also desirable, because that allows to give an Option to constructors of `@immutable` classes.

Reply:
However, Psalm will then demand that all the Closures you give to Option's methods must be marked as pure (there is a pseudo-type in Psalm for that, `pure-callable`, btw Psalm doesn't have similar type exactly for Closure). I am not sure we'd want to define our API allowing only side-effect-free Closures though - then `@psalm-suppress ImpureFunctionCall` will have to be used in `map()`, `unwrapOrElse()` and similar methods.

Reply:
functions taking closures cannot be pure, you can suppress the issue locally if desired, but suppressing in PSL is not an option as the callback could still carry side effects, using pure-callable is also not an option ( though, i was the one to introduce that type into psalm ^^" ).

but i agree on immutable/covariant 👍

Reply:
purity in psalm is insanity: https://github.com/vimeo/psalm/issues/8212, as of now we can't declare option as immutable.

Reply:
@azjezz hmm, in Psalm issue you wrote you can instead mark Option as `@readonly`. But that will not allow using it inside of `@immutable` classes.

Reply:
Hm 🤔 yes, the only way to actually have `Option` as immutable is to mark all Closure arguments are pure functions, which i'm not infavor of, since now you are *forced* to only use pure functions when using `Option` ( e.g: this is not possible anymore `$user->map(fn($u) => $repo->getUserOrgnizations($u));` )

Reply:
You can add impure closures as arguments if you suppress all function calls inside the option calls. It sucks, but it works...
We do that in our internal codebase as well

Reply:
the issue is end users would be forced to use pure functions only with option, which is not something i like to do

Reply:
( same reason why `Vec\map` and friends are not pure )

Reply:
No I meant you can use regular closures as param if you suppress the impure function call inside the option class.

That way you can provide impure functions to  the option method, even though it is immutable.

Not behind the computer for a couple of days, so it's hard to type an example on my phone  🙂

Reply:
Allowing non-pure closures also has its drawbacks, consider a mutable entity `User` with a method `setName()`:
```php
/** @var Option<User> */
$option->map(fn(User $user) => $user->setName('New name'));
```
which is not exactly what you'd expect from `map()`. However, restricting to only pure closures in PHP ecosystem is even worse. Therefore I guess, the best compromise is to suppress ImpureFunctionCall as @veewee suggests.

Reply:
https://psalm.dev/r/066c2e385b

This took me ages to write on a phone 🤣
But as you can see, usercode can use impure  closures on immutable objects this way.

Reply:
that's not good, now psalm reports this as okay, while it's not: https://psalm.dev/r/52bae7b6bf

x($f) is impure, because $f is impure.

Reply:
True ... That's the downside.
So it needs to be fixed in / added to psalm before it can be annotated as immutable-ish

Reply:
I am even thinking, maybe we should introduce `Option` accepting only pure callables, and another `MutableOption` which template is not covariant, and it accepts arbitrary callables?

Reply:
Adding a mutable version would be a step backward imo.
Maybe not marking it as immutable for psalm is ok for now? Would that have any downsides?
The only thing I can think of is that you won't be able to use it in pure functions, unless we mark the static constructors as pure maybe?

Reply:
psalm does not allow non-immutable classes to be `@template-covariant`

Reply:
@someniatko it does! however, it won't allow them it accept callables that use the template as arguments :(

Reply:
> functions taking closures cannot be pure, you can suppress the issue locally if desired, but suppressing in PSL is not an option as the callback could still carry side effects

Do we even care if it carries side effects or not? As soon as it should not mutate the Option itself?